### DISCUSSION TITLE
Calculate snapshot_weigtings when using datetimeindex

### DISCUSSION THREAD

**Discussion by martinhjel** [PyPSA/PyPSA#639] [Ideas]
When calling [`.set_snapshot(value)`](https://github.com/PyPSA/PyPSA/blob/master/pypsa/components.py#L457) the default snapshot_weights are 1. If the `value` is of type `pd.DatetimeIndex`, suggest calculating the hours between each time step and use that instead: 

```
  if isinstance(value, pd.DatetimeIndex):
      hours_per_step=(value.to_series()
          .diff(periods=1)
          .shift(-1)
          .ffill() # fill last value by assuming same as the one before
          .apply(lambda x: x.total_seconds()/3600)
      )
      self.snapshot_weightings = self.snapshot_weightings.apply(lambda x: hours_per_step)
```

**Comment by martinhjel** [comment]
Or at least throwing a warning that the snapshot weightings and time steps do not align.

**Comment by fneum** [comment]
@martinhjel, this is a very nice idea! But it may not be suitable for all applications of PyPSA. Consider, for instance, a system without storage where the snapshots are only related by the investments and not by consecutive hours. The snapshots can also give a snapshot a weight unrelated to time (e.g. for stochastic optimisation).

Nevertheless, the function you outline would make a very fine utility function in PyPSA (e.g. something like `set_snapshot_weightings_from_timedelta()`). Would you like to prepare a pull request?

**Comment by fneum** [comment]
A good position for this would be around the `set_snapshots` functions.

https://github.com/PyPSA/PyPSA/blob/master/pypsa/components.py

**Comment by FabianHofmann** [comment]
Another option would be to add a keyword argument `weightings_from_timedelta` to the `n.set_snapshots` function which is per default `False`, and can be set to `True`.

**Comment by martinhjel** [comment]
Suggested a PR here: #641
